File: src/main/java/File.java
import java.time.Instant;

class File {
    short attributes;
    byte[] content;

    long ctimesec;
    int ctimensec;

    long mtimesec;
    int mtimensec;

    long atimesec;
    int atimensec;

    int uid;
    int gid;

    public File(byte[] content, short attributes) {
        this.content = content;
        this.attributes = attributes;

        Instant now = Instant.now();
        this.ctimesec = now.getEpochSecond();
        this.ctimensec = now.getNano();
        this.mtimesec = now.getEpochSecond();
        this.mtimensec = now.getNano();
        this.atimesec = now.getEpochSecond();
        this.atimensec = now.getNano();

        this.uid = 501;              //FIXME
        this.gid = 20;               //FIXME
    }

}

File: src/main/scala/org/acef304/yafs/File.scala
package org.acef304.yafs

import java.time.Instant

case class Time(timesec: Long, timensec: Int)

object Time {
  def ofNow(): Time = {
    val now = Instant.now
    Time(now.getEpochSecond, now.getNano)
  }
}

case class File(content: FileContent, attributes: Short, ctime: Time, atime: Time, mtime: Time, uid: Int, gid: Int) {
  def setAtime(time: Instant = Instant.now()): File =
    this.copy(atime = Time(time.getEpochSecond, time.getNano))

  def setCtime(time: Instant = Instant.now()): File =
    this.copy(ctime = Time(time.getEpochSecond, time.getNano))

  def setMtime(time: Instant = Instant.now()): File =
    this.copy(mtime = Time(time.getEpochSecond, time.getNano))
}

object File {
  val defaultAttribute: Short = (7 * 8 * 8 + 7 * 8 + 7).toShort //means 0777
  val defaultUid = 501
  val defaultGid = 20

  def withContent(content: Array[Byte]): File = {
    File(
      content = FileContent.fromByteArray(content),
      attributes = defaultAttribute,
      ctime = Time.ofNow(),
      atime = Time.ofNow(),
      mtime = Time.ofNow(),
      uid = defaultUid,
      gid = defaultGid
    )
  }
}
